### PR TITLE
CDAP-15020 Implement tests for CDC GoldenGate Source plugin

### PR DESCRIPTION
**Pull request by mlozbin-cybervisiontech**
In scope of this PR:
- implemented unit and integration tests for Golden Gate Kafka Source plugin
- added Oracle and GoldenGate services to docker-compose
- added guide how to configure local Golden Gate with Oracle for tests
- fixed typo in Golden Gate Kafka Source plugin reference docs
- fixed exception in Golden Gate Kafka Source plugin

### CODE REVIEW

**Comment by mlozbin-cybervisiontech** [comment]
@jayfjin could you, please, take a look

File: docker-compose/cdc-env/Oracle/createDB.sh [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+# LICENSE UPL 1.0
+#
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (jayfjin) [inline docker-compose/cdc-env/Oracle/createDB.sh:4]: Does this have the correct copyright/license for us to just copy the code like this?

File: docker-compose/cdc-env/Oracle/dbca.rsp.tmpl [hunk 2]
```diff
@@ -0,0 +1,204 @@
+##############################################################################
+##                                                                          ##
+##                            DBCA response file                            ##
+##                            ------------------                            ##
+## Copyright(c) Oracle Corporation 1998,2017. All rights reserved.          ##
```
Comment (jayfjin) [inline docker-compose/cdc-env/Oracle/dbca.rsp.tmpl:5]: same here

File: src/main/java/co/cask/cdc/plugins/common/OperationType.java [hunk 3]
```diff
@@ -20,5 +20,17 @@
...
+  public static OperationType fromShortName(String name) {
+    switch (name.toUpperCase()) {
+      case "I":
+        return INSERT;
+      case "U":
+        return UPDATE;
+      case "D":
+        return DELETE;
+    }
+    throw new IllegalArgumentException(String.format("Unknown change operation '%s'", name));
```
Comment (jayfjin) [inline src/main/java/co/cask/cdc/plugins/common/OperationType.java]: move to default option for switch

File: src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@RunWith(Enclosed.class)
+public class GoldenGateKafkaPluginIntegrationTest {
```
Comment (jayfjin) [inline src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java:64]: do any configs/flags need to be set to run this test?

File: pom.xml [hunk 5]
```diff
@@ -91,7 +91,7 @@
     <netty.version>4.1.16.Final</netty.version>
     <junit.version>4.11</junit.version>
     <!-- properties for script build step that creates the config files for the artifacts -->
-    <app.parents>system:cdap-data-streams[4.1.0-SNAPSHOT,10.0.0)</app.parents>
+    <app.parents>system:cdap-data-streams[4.0.0,10.0.0)</app.parents>
```
Comment (jayfjin) [inline pom.xml:94]: why the change from 4.1 to 4.0?

File: README.md [hunk 6]
```diff
@@ -49,8 +49,19 @@ To use **remote environment** you may configure the following system properties:
 * **test.bigtable.instance** - Bigtable Instance ID. Default: null.
 * **test.bigtable.serviceFilePath** - Path on the local file system of the service account key used for
   authorization. Default: lookup from local environment.
+* **test.oracle-db.host** - Oracle DB host. Default: localhost.
+* **test.oracle-db.port** - Oracle DB port. Default: 1521.
+* **test.oracle-db.service** - Oracle DB service name. Default: XE.
+* **test.oracle-db.username** - Oracle DB username. Default: trans_user.
```
Comment (jayfjin) [inline README.md:55]: any reason why the default is "trans_user"?

File: docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm [hunk 7]
```diff
@@ -0,0 +1,7 @@
+REPLICAT rconf
```
Comment (jayfjin) [inline docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm:1]: just curious, what does this file do?

File: docker-compose/cdc-env/GoldenGate/dirprm/ext1.prm [hunk 8]
```diff
@@ -0,0 +1,4 @@
+EXTRACT ext1
```
Comment (jayfjin) [inline docker-compose/cdc-env/GoldenGate/dirprm/ext1.prm:1]: wondering what this does as well

File: docker-compose/cdc-env/Oracle/createDB.sh (hunk 1 above)
Comment (mlozbin-cybervisiontech) [reply docker-compose/cdc-env/Oracle/createDB.sh:4]: As I understood we can modify and use content if case we do not change license header.

Sources:
- https://www.oracle.com/technetwork/licenses/upl-license-2927578.html
- https://tldrlegal.com/license/universal-permissive-license-1.0-(upl-1.0)

File: docker-compose/cdc-env/Oracle/dbca.rsp.tmpl (hunk 2 above)
Comment (mlozbin-cybervisiontech) [reply docker-compose/cdc-env/Oracle/dbca.rsp.tmpl:5]: This header does not contain a reference to license, so I think we can use it but I'm not sure.

File: src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java (hunk 4 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java:64]: Everything is described in README.md. Should I copy some description to this class?

File: pom.xml (hunk 5 above)
Comment (mlozbin-cybervisiontech) [reply pom.xml:94]: Actually that was not changed. Previously this property was not used in pom file. I've copied value from previous pom.

File: README.md (hunk 6 above)
Comment (mlozbin-cybervisiontech) [reply README.md:55]: this value is used in default local environment (from docker-compose)

File: docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm (hunk 7 above)
Comment (mlozbin-cybervisiontech) [reply docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm:1]: this file is a set of configuration params for GoldenGate's REPLICAT.

File: docker-compose/cdc-env/GoldenGate/dirprm/ext1.prm (hunk 8 above)
Comment (mlozbin-cybervisiontech) [reply docker-compose/cdc-env/GoldenGate/dirprm/ext1.prm:1]: this file is a set of configuration params for GoldenGate's EXTRACT.

File: src/main/java/co/cask/cdc/plugins/common/OperationType.java (hunk 3 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/co/cask/cdc/plugins/common/OperationType.java]: sure, will move

File: docker-compose/cdc-env/Oracle/createDB.sh (hunk 1 above)
Comment (jayfjin) [reply docker-compose/cdc-env/Oracle/createDB.sh:4]: sg, thanks for checking

File: src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java (hunk 4 above)
Comment (jayfjin) [reply src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java:64]: Can we add the exact command you ran to run the test in the readme?

File: docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm (hunk 7 above)
Comment (jayfjin) [reply docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm:1]: Let's add a quick description to these files in the readme

File: src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java (hunk 4 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java:64]: the command to run all tests is in readme file. Should I also add a manual how to run single test using maven? This task is pretty common and described in maven documentation. Or do you mean something else?

Comment (jayfjin) [reply src/test/java/co/cask/cdc/plugins/integration/source/GoldenGateKafkaPluginIntegrationTest.java:64]: oh ok didn't see that

File: docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm (hunk 7 above)
Comment (jayfjin) [reply docker-compose/cdc-env/GoldenGate-Bigdata/dirprm/rconf.prm:1]: will merge once this is done

**Comment by mlozbin-cybervisiontech** [comment]
@jayfjin pushed commit that addresses your latest comments
